src/lib/optimizer/strategy/or_to_union_rule.cpp

```diff
@@ -0,0 +1,34 @@
+#include <expression/expression_utils.hpp>
+
+#include "expression/logical_expression.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "or_to_union_rule.hpp"
+
+namespace opossum {
+
+std::string OrToUnionRule::name() const { return "Or to Union Rule"; }
+
+void OrToUnionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  if (node->type == LQPNodeType::Predicate) {
```

Comment:
Use visit_lqp to avoid redundant traversal of subplans under diamond shapes

Reply:
(when using visit_lqp you need to only collect the nodes to be replaced in the lambda and then to the replacing outside of visit_lqp)

Reply:
You could also consider moving this to the top level of `ExpressionReductionRule::apply_to` where we are already visiting the LQP.

Reply:
ExpressionReductionRule is for Expressions that get rewritten into expressions, not for expressions that are rewritten into operators imo

Reply:
@mrzzzrm is right. The ExpressionReductionRule does not change the nodes of the LQP.
(Another aspect is that the DisjunctionToUnionRule is the OR equivalent of the PredicateSplitUpRule (AND). If one rule becomes part of the ExpressionReductionRule, the other one should as well.)